### DISCUSSION TITLE
[Help] How to resolve complex nested structures using winnow

### DISCUSSION THREAD

**Discussion by yyy33** [winnow-rs/winnow#371] [Q&A]
Here is the document I want to parse, including ` keyN ` and two parentheses. I use `()` to say that the elements are ordered, I use ` {} ` to say that the elements are disordered, and `()` and `()` {} ` can be nested with each other. How can I use winnow to parse this structure

````

(key1 key3  key2)
{key1 key3  key2}
(key9 key3 {key2 key4 key7 (key11 key20 (key11 key12) key30)})

````

Below is the data structure that I defined in rust

````rust

struct Bracket {
    units: Vec<BracketUnit>,
    order: bool,
}

enum BracketUnit {
    Base(Key),
    Bracket(Bracket),
}

struct Key {
    seq: i8,
}

````

**Comment by epage** [comment]
Is there a specific problem you ran into in trying to write your parser with winnow? Thatd help in knowing how to best give next steps in you writing your parser.

If you haven't checked out the json parsers we have under Special Topics in the docs, I'd recommend those as they show delimited, wrapped values.

**Reply by yyy33** [reply]
Thanks, I think I know how to get started.
